physics/cu_unified_deep.F90

```diff
@@ -0,0 +1,6051 @@
+              ierr,z1,                                              &
+              itf,ktf,                                              &
+              its,ite, kts,kte                       )
+
+   implicit none
+
+     integer                                                        &
+        ,intent (in   )                   ::                        &
+        itf,ktf,                                                    &
+        its,ite, kts,kte
+  !
+     real(kind=kind_phys),    dimension (its:ite,kts:kte)                        &
+        ,intent (in   )                   ::                     &
+        qes,q,he,hes,z,p,t
+!$acc declare copyin(qes,q,he,hes,z,p,t)
+     real(kind=kind_phys),    dimension (its:ite,kts:kte)                        &
+        ,intent (out  )                   ::                     &
+        qes_cup,q_cup,he_cup,hes_cup,z_cup,p_cup,gamma_cup,t_cup
+!$acc declare copyout(qes_cup,q_cup,he_cup,hes_cup,z_cup,p_cup,gamma_cup,t_cup)
+     real(kind=kind_phys),    dimension (its:ite)                                &
+        ,intent (in   )                   ::                     &
+        psur,z1
+!$acc declare copyin(psur,z1)
+     integer, dimension (its:ite)                                &
+        ,intent (inout)                   ::                     &
+        ierr
+!$acc declare copy(ierr)
+!
+!  local variables in this routine
+!
+
+     integer                              ::                     &
+       i,k
+
+!$acc kernels
+      do k=kts,ktf
+      do i=its,itf
+        qes_cup(i,k)=0.
+        q_cup(i,k)=0.
+        hes_cup(i,k)=0.
+        he_cup(i,k)=0.
+        z_cup(i,k)=0.
+        p_cup(i,k)=0.
+        t_cup(i,k)=0.
+        gamma_cup(i,k)=0.
+      enddo
+      enddo
+      do k=kts+1,ktf
+      do i=its,itf
+        if(ierr(i).eq.0)then
+        qes_cup(i,k)=.5*(qes(i,k-1)+qes(i,k))
+        q_cup(i,k)=.5*(q(i,k-1)+q(i,k))
+        hes_cup(i,k)=.5*(hes(i,k-1)+hes(i,k))
+        he_cup(i,k)=.5*(he(i,k-1)+he(i,k))
+        if(he_cup(i,k).gt.hes_cup(i,k))he_cup(i,k)=hes_cup(i,k)
+        z_cup(i,k)=.5*(z(i,k-1)+z(i,k))
+        p_cup(i,k)=.5*(p(i,k-1)+p(i,k))
+        t_cup(i,k)=.5*(t(i,k-1)+t(i,k))
+        gamma_cup(i,k)=(xlv/cp)*(xlv/(r_v*t_cup(i,k) &
+                       *t_cup(i,k)))*qes_cup(i,k)
+        endif
+      enddo
+      enddo
+      do i=its,itf
+        if(ierr(i).eq.0)then
+        qes_cup(i,1)=qes(i,1)
+        q_cup(i,1)=q(i,1)
+!       hes_cup(i,1)=hes(i,1)
+!       he_cup(i,1)=he(i,1)
+        hes_cup(i,1)=9.81*z1(i)+1004.*t(i,1)+2.5e6*qes(i,1)
+        he_cup(i,1)=9.81*z1(i)+1004.*t(i,1)+2.5e6*q(i,1)
+        z_cup(i,1)=.5*(z(i,1)+z1(i))
+        p_cup(i,1)=.5*(p(i,1)+psur(i))
+        z_cup(i,1)=z1(i)
+        p_cup(i,1)=psur(i)
+        t_cup(i,1)=t(i,1)
+        gamma_cup(i,1)=xlv/cp*(xlv/(r_v*t_cup(i,1) &
+                       *t_cup(i,1)))*qes_cup(i,1)
+        endif
+      enddo
+!$acc end kernels
+   end subroutine cup_env_clev
+
```

Comment:
@dustinswales @grantfirl @Qingfu-Liu the addition of the prognostic closure in this subroutine is new